File: 0075-sort-colors/0075-sort-colors.java
class Solution {
    public void sortColors(int[] nums) {
        // Initialize pointers for the current element (currIndex),
        // the last position of 0 (lastZeroIndex) and the first position of 2 (firstTwoIndex)
        int lastZeroIndex = -1;
        int firstTwoIndex = nums.length;
        int currIndex = 0;

        // Process elements until currIndex reaches firstTwoIndex
        while (currIndex < firstTwoIndex) {
            if (nums[currIndex] == 0) {
                // If the current element is 0, swap it to the position after the last 0 we found
                swap(nums, ++lastZeroIndex, currIndex++);
            } else if (nums[currIndex] == 2) {
                // If the current element is 2, swap it with the element at the position
                // just before the first 2 we found
                swap(nums, --firstTwoIndex, currIndex);
            } else { 
                // If the current element is 1, just move to the next element
                ++currIndex;
            }
        }
    }

    // Helper method to swap two elements in an array
    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

File: 0075-sort-colors/0075-sort-colors.swift
class Solution {
    func sortColors(_ nums: inout [Int]) {
        let n = nums.count
        var lo = 0, hi = n - 1, mid = 0
        
        while mid <= hi {
            if nums[mid] == 0 {
                nums.swapAt(lo, mid)
                lo += 1
                mid += 1
            } else if nums[mid] == 1 {
                mid += 1
            } else { // nums[mid] == 2
                nums.swapAt(hi, mid)
                hi -= 1
            }
        }
    }
}